qiita_db/test/test_util.py

```diff
@@ -421,6 +422,68 @@ def test_purge_filepaths(self):
         self.assertTrue(exists(fp1))
         self.assertFalse(exists(fp_exp_id))
 
+    def test_purge_filepaths_null_cols(self):
+        """Tests the nulls vs not in issue in purge_filepaths"""
+        # For more details about the issue:
+        # http://www.depesz.com/2008/08/13/nulls-vs-not-in/
+        # In the current set up, the only place where we can actually have a
+        # null value in a filepath id is in the reference table. Add a new
+        # reference without tree and taxonomy:
+        fd, seqs_fp = mkstemp(suffix="_seqs.fna")
+        close(fd)
+        ref = Reference.create("null_db", "13_2", seqs_fp)
+        self.files_to_remove.append(ref.sequence_fp)
+        # Add a new filepath to the database
+        fd, fp = mkstemp()
+        close(fd)
+        exp_id = 1 + self.conn_handler.execute_fetchone(
+            "SELECT count(1) FROM qiita.filepath")[0]
+        fp_id = self.conn_handler.execute_fetchone(
+            "INSERT INTO qiita.filepath "
+            "(filepath, filepath_type_id, checksum, checksum_algorithm_id) "
+            "VALUES (%s, %s, %s, %s) RETURNING filepath_id", (fp, 1, "", 1))[0]
+        self.assertEqual(fp_id, exp_id)
+
+        # Connect the just added filepath to a raw data
+        self.conn_handler.execute(
+            "INSERT INTO qiita.raw_filepath (raw_data_id, filepath_id) VALUES"
+            "(%s, %s)", (1, exp_id))
+
+        # Get the filepaths so we can test if they've been removed or not
+        sql_fp = "SELECT filepath FROM qiita.filepath WHERE filepath_id=%s"
+        fp1 = self.conn_handler.execute_fetchone(sql_fp, (1,))[0]
+        fp1 = join(get_db_files_base_dir(), fp1)
+
+        # Make sure that the file exists - specially for travis
+        with open(fp1, 'w') as f:
+            f.write('\n')
+
+        fp_exp_id = self.conn_handler.execute_fetchone(sql_fp, (exp_id,))[0]
+        fp_exp_id = join(get_db_files_base_dir(), fp_exp_id)
+
+        purge_filepaths(self.conn_handler)
+
+        # Check that the files still exist
+        self.assertTrue(exists(fp1))
+        self.assertTrue(exists(fp_exp_id))
+
+        # Unlink the filepath from the raw data
+        self.conn_handler.execute(
+            "DELETE FROM qiita.raw_filepath WHERE filepath_id=%s", (exp_id,))
+
+        # Only filepath 16 should be removed
+        sql = ("SELECT filepath_id FROM qiita.filepath ORDER BY "
+               "filepath_id")
+        exp_ids = [i[0] for i in self.conn_handler.execute_fetchall(sql)]
+        exp_ids.pop()
+        purge_filepaths(self.conn_handler)
+        obs_ids = [i[0] for i in self.conn_handler.execute_fetchall(sql)]
+        self.assertEqual(obs_ids, exp_ids)
+
+        # Check that only the file for the removed filepath has been removed
```

Comment:
Hmmm if nobody else feels like this should be explicitly tested, then
I'm ok with the current state.

On (Feb-13-15|13:22), josenavas wrote:
> > +        self.assertTrue(exists(fp_exp_id))
> > +
> > +        # Unlink the filepath from the raw data
> > +        self.conn_handler.execute(
> > +            "DELETE FROM qiita.raw_filepath WHERE filepath_id=%s", (exp_id,))
> > +
> > +        # Only filepath 16 should be removed
> > +        sql = ("SELECT filepath_id FROM qiita.filepath ORDER BY "
> > +               "filepath_id")
> > +        exp_ids = [i[0] for i in self.conn_handler.execute_fetchall(sql)]
> > +        exp_ids.pop()
> > +        purge_filepaths(self.conn_handler)
> > +        obs_ids = [i[0] for i in self.conn_handler.execute_fetchall(sql)]
> > +        self.assertEqual(obs_ids, exp_ids)
> > +
> > +        # Check that only the file for the removed filepath has been removed
> 
> We are testing that only one of the other ones still exists.
> 
> Fun fact: the filepaths that are present in the database, they actually do not exist. Which means that if `purge_filepaths` try to remove them, it will fail, which will show up here. So, it is actually testing that is correctly working just by the fact that is not trying to remove those other files :-)
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/biocore/qiita/pull/837/files#r24697777